ethcore/src/migrations/account_bloom.rs

```diff
@@ -0,0 +1,93 @@
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Bloom upgrade
+
+use client::{DB_COL_EXTRA, DB_COL_HEADERS, DB_NO_OF_COLUMNS, DB_COL_STATE};
+use state_db::{ACCOUNT_BLOOM_SPACE, DEFAULT_ACCOUNT_PRESET, StateDB};
+use util::trie::TrieDB;
+use views::HeaderView;
+use bloomfilter::Bloom;
+use util::migration::Error;
+use util::journaldb;
+use util::{H256, FixedHash, Hashable, BytesConvertable};
+use util::{Database, DatabaseConfig, DBTransaction, CompactionProfile};
+use std::path::Path;
+
+/// Account bloom upgrade routine. If bloom already present, does nothing.
+/// If database empty (no best block), does nothing.
+/// Can be called on upgraded database with no issues (will do nothing).
+pub fn upgrade_account_bloom(db_path: &Path) -> Result<(), Error> {
+	let path = try!(db_path.to_str().ok_or(Error::MigrationImpossible));
+	let source = try!(Database::open(&DatabaseConfig {
+		max_open_files: 64,
+		cache_size: None,
+		compaction: CompactionProfile::default(),
+		columns: DB_NO_OF_COLUMNS,
+		wal: true,
+	}, path));
+
+	let best_block_hash = match try!(source.get(DB_COL_EXTRA, b"best")) {
+		// no migration needed
+		None => { return Ok(()); }
+		Some(hash) => hash,
+	};
+	let best_block_header = match try!(source.get(DB_COL_HEADERS, &best_block_hash)) {
+		// no best block, nothing to do
+		None => {
+			trace!(target: "migration", "No best block, skipping");
+			return Ok(())
+		},
+		Some(x) => x,
+	};
+	let state_root = HeaderView::new(&best_block_header).state_root();
+
+	if StateDB::check_bloom_exists(&source) {
+		// bloom already exists, nothing to do
+		trace!(target: "migration", "Bloom already present, skipping");
+		return Ok(())
+	}
+
+	println!("Adding accounts bloom (one-time upgrade). Please don't close parity.");
+	let db = ::std::sync::Arc::new(source);
+	let bloom_journal = {
+		let mut bloom = Bloom::new(ACCOUNT_BLOOM_SPACE, DEFAULT_ACCOUNT_PRESET);
+		// no difference what algorithm is passed, since there will be no writes
+		let state_db = journaldb::new(
+			db.clone(),
+			journaldb::Algorithm::OverlayRecent,
+			DB_COL_STATE);
+		let account_trie = try!(TrieDB::new(state_db.as_hashdb(), &state_root).map_err(|e| Error::Custom(format!("Cannot open trie: {:?}", e))));
+		for (ref account_key, _) in account_trie.iter() {
+			let account_key_hash = H256::from_slice(&account_key);
+			bloom.set(account_key_hash.sha3().as_slice());
+		}
+
+		bloom.drain_journal()
+	};
+
+	trace!(target: "migration", "Generated {} bloom updates", bloom_journal.entries.len());
+
+	let batch = DBTransaction::new(&db);
```

Comment:
Unsure if this can or not happend concurrently
would be nice if @rphmeier review this